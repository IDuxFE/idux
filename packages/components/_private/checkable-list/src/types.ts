/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
