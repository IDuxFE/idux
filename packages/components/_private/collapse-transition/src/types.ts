/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
