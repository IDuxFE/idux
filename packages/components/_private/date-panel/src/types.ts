/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
