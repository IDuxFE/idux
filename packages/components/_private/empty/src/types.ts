/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { EmptyProps as IxEmptyProps } from '@idux/components/empty'

export interface EmptyProps {
  empty?: string | IxEmptyProps
}
