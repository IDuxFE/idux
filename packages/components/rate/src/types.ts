/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { FormSize } from '@idux/components/form'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const rateProps = {
  value: [Number, String] as PropType<number | string>,
  allowHalf: {
    type: Boolean,
    default: undefined,
  },
  clearable: {
    type: Boolean,
    default: undefined,
  },
  count: [Number, String] as PropType<number | string>,
  disabled: {
    type: Boolean,
    default: false,
  },
  icon: String,
  tooltips: {
    type: Array as PropType<string[]>,
    default: (): string[] => [],
  },
  size: String as PropType<FormSize>,
  color: {
    type: String,
    default: undefined,
  },

  // events
  'onUpdate:value': [Function, Array] as PropType<MaybeArray<(value: number) => void>>,
  onChange: [Function, Array] as PropType<MaybeArray<(value: number) => void>>,
  onBlur: [Function, Array] as PropType<MaybeArray<(evt: FocusEvent) => void>>,
  onFocus: [Function, Array] as PropType<MaybeArray<(evt: FocusEvent) => void>>,
  onKeyDown: [Function, Array] as PropType<MaybeArray<(evt: KeyboardEvent) => void>>,
} as const

export type RateProps = ExtractInnerPropTypes<typeof rateProps>
export type RatePublicProps = ExtractPublicPropTypes<typeof rateProps>
export interface RateBindings {
  blur: () => void
  focus: (options?: FocusOptions) => void
}
export type RateComponent = DefineComponent<Omit<HTMLAttributes, keyof RatePublicProps> & RatePublicProps, RateBindings>
export type RateInstance = InstanceType<DefineComponent<RateProps, RateBindings>>

// private
export const rateItemProps = {
  count: {
    type: Number,
    required: true,
  },
  disabled: {
    type: Boolean,
    required: true,
  },
  focused: {
    type: Boolean,
    required: true,
  },
  index: {
    type: Number,
    required: true,
  },
  prefixCls: {
    type: String,
    required: true,
  },
  tooltip: String,
  value: {
    type: Number,
    required: true,
  },
  color: String,

  // events
  onClick: {
    type: Function as PropType<(evt: MouseEvent, element: HTMLElement, index: number) => void>,
    required: true,
  },
  onMouseMove: {
    type: Function as PropType<(evt: MouseEvent, element: HTMLElement, index: number) => void>,
    required: true,
  },
} as const
