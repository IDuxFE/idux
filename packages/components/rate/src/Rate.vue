<template>
  <div class="ix-rate-wrap">
    <div
      v-for="(item, index) in rateCount"
      :key="index"
      :ref="
        el => {
          if (el) starRefs[index] = el
        }
      "
      class="ix-rate-item"
      :title="getTooltip(index)"
      :style="{ cursor: disabled ? 'auto' : 'pointer' }"
      @mousemove="handleMouseEnter(item, $event)"
      @mouseleave="handleMouseLeave(item)"
      @click="handleClick(item, $event)"
    >
      <div class="ix-rate-full">
        <ix-icon :name="rateIcon" class="ix-rate-iconfont-main" :class="getIconClass(item)" />
      </div>

      <div v-if="showDecimalIcon(item)" class="ix-rate-half">
        <ix-icon :name="rateIcon" class="ix-rate-iconfont-main ix-rate-half-icon" />
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, onBeforeUpdate, ref, watchEffect } from 'vue'
import { isArray, toNumber } from '@idux/cdk/utils'
import { IxIcon } from '@idux/components/icon'
import { useGlobalConfig } from '@idux/components/config'
import { rateProps } from './types'

const HALF = 2

export default defineComponent({
  name: 'IxRate',
  components: { IxIcon },
  props: rateProps,
  emits: ['update:value', 'change'],
  setup(props, { emit }) {
    const score = ref(props.value)
    const touchHalf = ref(false)
    const starRefs = ref([])

    const rateGlobalConfig = useGlobalConfig('rate')
    const rateCount = computed(() => toNumber(props.count ?? rateGlobalConfig.count))
    const rateIcon = computed(() => props.icon ?? rateGlobalConfig.icon)
    const allowHalf = computed(() => props.allowHalf ?? rateGlobalConfig.allowHalf)
    const allowClear = computed(() => props.allowClear ?? rateGlobalConfig.allowClear)

    onBeforeUpdate(() => {
      starRefs.value = []
    })

    watchEffect(() => {
      const value = toNumber(props.value)
      score.value = value
      touchHalf.value = value !== Math.floor(value)
    })

    function getIconClass(item: number) {
      return item <= score.value ? 'ix-rate-highlight' : 'ix-rate-normal'
    }

    function showDecimalIcon(item: number) {
      return allowHalf.value && item - 0.5 <= score.value && item > score.value
    }

    function getTooltip(index: number) {
      if (!props.tooltips || !isArray(props.tooltips)) {
        return
      }

      return props.tooltips[index] || ''
    }

    function handleMouseEnter(item: number, e: MouseEvent) {
      if (props.disabled) {
        return
      }

      if (allowHalf.value) {
        updateTouchHalf(item - 1, e)
        score.value = touchHalf.value ? item - 0.5 : item
      } else {
        score.value = item
      }
    }

    function updateTouchHalf(index: number, e: MouseEvent) {
      const target = starRefs.value[index] as HTMLElement

      if (e.offsetX <= target.clientWidth / HALF) {
        touchHalf.value = true
        return
      }

      touchHalf.value = false
    }

    function handleMouseLeave() {
      if (props.disabled) {
        return
      }

      const value = toNumber(props.value)

      if (allowHalf.value) {
        touchHalf.value = value !== Math.floor(value)
      }
      score.value = value
    }

    function handleClick() {
      if (props.disabled) {
        return
      }

      let clearValue = false
      let emitValue = 0
      const beforeValue = toNumber(props.value)
      const currentScore = toNumber(score.value)

      if (allowClear.value) {
        clearValue = beforeValue === currentScore
      }

      emitValue = clearValue ? 0 : currentScore

      if (emitValue !== beforeValue) {
        emit('update:value', emitValue)
        emit('change', emitValue)
      }
    }

    return {
      rateCount,
      rateIcon,
      getTooltip,
      showDecimalIcon,
      handleMouseEnter,
      handleMouseLeave,
      handleClick,
      getIconClass,
      starRefs,
    }
  },
})
</script>
