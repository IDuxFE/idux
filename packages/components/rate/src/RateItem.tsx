/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, ref } from 'vue'

import { IxTooltip } from '@idux/components/tooltip'

import { rateItemProps } from './types'

export default defineComponent({
  props: rateItemProps,
  setup(props, { slots }) {
    const liRef = ref<HTMLElement>()

    const handleClick = (evt: MouseEvent) => props.onClick(evt, liRef.value!, props.index)

    const handleKeyDown = (evt: KeyboardEvent) => {
      if (evt.code === 'Enter') {
        handleClick(evt as unknown as MouseEvent)
      }
    }

    const handleMouseMove = (evt: MouseEvent) => props.onMouseMove(evt, liRef.value!, props.index)

    const classes = computed(() => {
      const { focused, index, prefixCls, value } = props
      const itemValue = index + 1

      const isHalf = value + 0.5 >= itemValue && value < itemValue
      const isFull = !isHalf && itemValue <= value
      const isZero = !isFull && !isHalf
      return {
        [prefixCls]: true,
        [`${prefixCls}-focused`]: focused && (isHalf || value === itemValue || (index === 0 && value === 0)),
        [`${prefixCls}-full`]: isFull,
        [`${prefixCls}-half`]: isHalf,
        [`${prefixCls}-zero`]: isZero,
      }
    })

    return () => {
      const { count, disabled, index, prefixCls, tooltip, value, color } = props

      const iconNode = slots.default!()
      const itemNode = (
        <li ref={liRef} class={classes.value} style={`color: ${color}`}>
          <span
            aria-checked={value > index}
            aria-posinset={index + 1}
            aria-setsize={count}
            role="radio"
            tabindex={disabled ? -1 : 0}
            onClick={disabled ? undefined : handleClick}
            onKeydown={disabled ? undefined : handleKeyDown}
            onMousemove={disabled ? undefined : handleMouseMove}
          >
            <span class={`${prefixCls}-first`}>{iconNode}</span>
            <span class={`${prefixCls}-second`}>{iconNode}</span>
          </span>
        </li>
      )

      return tooltip ? <IxTooltip title={tooltip}>{itemNode}</IxTooltip> : itemNode
    }
  },
})
