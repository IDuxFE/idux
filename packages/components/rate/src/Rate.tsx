/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass, ref } from 'vue'

import { callEmit, convertNumber } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { useFormAccessor, useFormElement, useFormSize } from '@idux/components/form'
import { IxIcon } from '@idux/components/icon'

import RateItem from './RateItem'
import { rateProps } from './types'

export default defineComponent({
  name: 'IxRate',
  props: rateProps,
  setup(props, { attrs, expose, slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-rate`)
    const config = useGlobalConfig('rate')
    const mergedSize = useFormSize(props, config)

    const { elementRef, focus, blur } = useFormElement()
    expose({ focus, blur })

    const accessor = useFormAccessor<number>()

    const countRef = computed(() => convertNumber(props.count ?? config.count))
    const iconRef = computed(() => props.icon ?? config.icon)
    const allowHalfRef = computed(() => props.allowHalf ?? config.allowHalf)
    const clearableRef = computed(() => props.clearable ?? config.clearable)
    const isDisabled = computed(() => accessor.disabled.value)
    const isFocused = ref(false)

    const hoverValue = ref<number>()

    const changeValue = (value: number) => {
      callEmit(props.onChange, value)
      accessor.setValue(value)
    }

    const calcValue = (evt: MouseEvent, element: HTMLElement, index: number) => {
      let value = index + 1
      if (allowHalfRef.value) {
        if (evt.offsetX < element.clientWidth / 2) {
          value -= 0.5
        }
      }
      return value
    }

    const handleItemClick = (evt: MouseEvent, element: HTMLElement, index: number) => {
      const currValue = accessor.valueRef.value
      const newValue = calcValue(evt, element, index)
      const isClear = clearableRef.value && currValue === newValue

      changeValue(isClear ? 0 : newValue)
    }

    const handleItemMouseMove = (evt: MouseEvent, element: HTMLElement, index: number) => {
      const newValue = calcValue(evt, element, index)
      hoverValue.value = newValue
    }

    const handleFocus = (evt: FocusEvent) => {
      isFocused.value = true
      callEmit(props.onFocus, evt)
    }

    const handleBlur = (evt: FocusEvent) => {
      isFocused.value = false
      callEmit(props.onBlur, evt)
      accessor.markAsBlurred()
    }

    const handleKeyDown = (evt: KeyboardEvent) => {
      const currValue = accessor.valueRef.value
      const currCount = countRef.value
      const currAllowHalf = allowHalfRef.value
      const eventCode = evt.code
      switch (eventCode) {
        case 'ArrowRight':
        case 'ArrowUp':
          {
            if (currValue < currCount) {
              changeValue(currValue + (currAllowHalf ? 0.5 : 1))
              evt.preventDefault()
            }
          }
          break
        case 'ArrowLeft':
        case 'ArrowDown':
          if (currValue > 0) {
            changeValue(currValue - (currAllowHalf ? 0.5 : 1))
            evt.preventDefault()
          }
          break
      }
    }

    const handleMouseLeave = () => {
      hoverValue.value = undefined
    }

    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-disabled`]: isDisabled.value,
        [`${prefixCls}-focused`]: isFocused.value,
        [`${prefixCls}-${mergedSize.value}`]: true,
      })
    })

    return () => {
      const count = countRef.value
      const disabled = isDisabled.value
      const focused = isFocused.value
      const itemValue = convertNumber(hoverValue.value ?? accessor.valueRef.value)
      const itemPrefixCls = `${mergedPrefixCls.value}-item`

      const { tooltips } = props
      const children = []
      for (let index = 0; index < count; index++) {
        children.push(
          <RateItem
            key={index}
            count={count}
            disabled={disabled}
            focused={focused}
            index={index}
            prefixCls={itemPrefixCls}
            tooltip={tooltips[index]}
            value={itemValue}
            onClick={handleItemClick}
            onMouseMove={handleItemMouseMove}
          >
            {slots.icon?.({ disabled, focused, index }) ?? <IxIcon name={iconRef.value} />}
          </RateItem>,
        )
      }

      return (
        <ul
          ref={elementRef}
          class={classes.value}
          tabindex={disabled ? -1 : (attrs.tabindex as number)}
          onFocus={disabled ? undefined : handleFocus}
          onBlur={disabled ? undefined : handleBlur}
          onKeydown={disabled ? undefined : handleKeyDown}
          onMouseleave={disabled ? undefined : handleMouseLeave}
          role="radiogroup"
        >
          {children}
        </ul>
      )
    }
  },
})
