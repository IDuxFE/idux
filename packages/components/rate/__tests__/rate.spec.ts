import { MountingOptions, VueWrapper, mount } from '@vue/test-utils'

import { renderWork } from '@tests'

import IxRate from '../src/Rate'
import { RateInstance, RateProps } from '../src/types'

// TODO rebuild
describe.skip('Rate', () => {
  const RateMount = (options?: MountingOptions<Partial<RateProps>>) => mount(IxRate, { ...options })
  let findNormalIcon: (wrapper: VueWrapper<RateInstance>) => number

  beforeEach(() => {
    findNormalIcon = (wrapper: VueWrapper<RateInstance>): number => {
      const res = wrapper.findAll('.ix-rate-iconfont-main').filter(item => {
        return item.classes().includes('ix-rate-normal')
      })

      return res.length
    }
  })

  renderWork(IxRate, { props: { value: 3 } })

  test('init hightlight as value', () => {
    const wrapper = RateMount({
      props: {
        value: 3,
      },
    })

    const normalIcon = findNormalIcon(wrapper)

    expect(normalIcon).toBe(2)
  })

  test('test count', () => {
    const wrapper = RateMount({
      props: {
        value: 3,
        count: 10,
      },
    })

    expect(wrapper.findAll('.ix-rate-full').length).toBe(10)
  })

  test('test tooltip', () => {
    const wrapper = RateMount({
      props: {
        value: 3,
        tooltips: ['1', '2', '3', '4', '5'],
      },
    })

    const titleList = wrapper.findAll('.ix-rate-item').filter(item => {
      return item.attributes()['title']
    })

    expect(titleList.length).toBe(5)
  })

  test('test click', async () => {
    const wrapper = RateMount({})

    await wrapper.setProps({ value: 3 })

    const thirdStar = wrapper.findAll('.ix-rate-item')[2].element as HTMLElement
    thirdStar.click()

    const normalIcon = findNormalIcon(wrapper)

    expect(normalIcon).toBe(2)
  })

  test('touch half and allow half', async () => {
    const wrapper = mount({
      template: `<IxRate v-model:value="value" allow-half ref="testRateRef" />`,
      data() {
        return {
          value: 4,
        }
      },
      components: { IxRate },
    })

    const vm = wrapper.vm
    const testHalfItem = wrapper.findAll('.ix-rate-item')[1]

    await testHalfItem.trigger('mousemove', { offsetX: 0 })
    await testHalfItem.trigger('click')
    await testHalfItem.trigger('mouseleave')

    expect(vm.value).toEqual(1.5)
  })

  test('disable', async () => {
    const wrapper = mount({
      template: `<IxRate v-model:value="value" disabled ref="testRateRef" />`,
      data() {
        return {
          value: 4,
        }
      },
      components: { IxRate },
    })

    const vm = wrapper.vm
    const testItem = wrapper.findAll('.ix-rate-item')[1]

    await testItem.trigger('mousemove')
    await testItem.trigger('click')
    await testItem.trigger('mouseleave')

    expect(vm.value).toEqual(4)
  })

  test('allow clear', async () => {
    const wrapper = mount({
      template: `<IxRate v-model:value="value" allow-clear ref="testRateRef" />`,
      data() {
        return {
          value: 2,
        }
      },
      components: { IxRate },
    })

    await wrapper.setProps({ value: 3 })

    const vm = wrapper.vm
    const testItem = wrapper.findAll('.ix-rate-item')[2]

    await testItem.trigger('mousemove', { offsetX: 26 })
    await testItem.trigger('click')
    await testItem.trigger('mouseleave')

    expect(vm.value).toEqual(0)
  })
})
