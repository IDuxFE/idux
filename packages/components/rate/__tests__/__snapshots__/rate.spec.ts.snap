// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rate.vue render work 1`] = `
"<div class=\\"ix-rate-wrap\\">
  <div class=\\"ix-rate-item\\" style=\\"cursor: pointer;\\">
    <div class=\\"ix-rate-full\\"><i class=\\"ix-icon ix-icon-star ix-rate-iconfont-main ix-rate-highlight\\" role=\\"img\\" aria-label=\\"star\\"></i></div>
    <!--v-if-->
  </div>
  <div class=\\"ix-rate-item\\" style=\\"cursor: pointer;\\">
    <div class=\\"ix-rate-full\\"><i class=\\"ix-icon ix-icon-star ix-rate-iconfont-main ix-rate-highlight\\" role=\\"img\\" aria-label=\\"star\\"></i></div>
    <!--v-if-->
  </div>
  <div class=\\"ix-rate-item\\" style=\\"cursor: pointer;\\">
    <div class=\\"ix-rate-full\\"><i class=\\"ix-icon ix-icon-star ix-rate-iconfont-main ix-rate-highlight\\" role=\\"img\\" aria-label=\\"star\\"></i></div>
    <!--v-if-->
  </div>
  <div class=\\"ix-rate-item\\" style=\\"cursor: pointer;\\">
    <div class=\\"ix-rate-full\\"><i class=\\"ix-icon ix-icon-star ix-rate-iconfont-main ix-rate-normal\\" role=\\"img\\" aria-label=\\"star\\"></i></div>
    <!--v-if-->
  </div>
  <div class=\\"ix-rate-item\\" style=\\"cursor: pointer;\\">
    <div class=\\"ix-rate-full\\"><i class=\\"ix-icon ix-icon-star ix-rate-iconfont-main ix-rate-normal\\" role=\\"img\\" aria-label=\\"star\\"></i></div>
    <!--v-if-->
  </div>
</div>"
`;
